subsys/testsuite/ztest/src/ztest_shell.c

```diff
@@ -0,0 +1,416 @@
+{
+	ARG_UNUSED(sh);
+	ARG_UNUSED(opts);
+
+	return 0;
+}
+
+static int cmd_ztest(const struct shell *sh, size_t argc, char *const *argv)
+{
+	int ret;
+
+	_state = (struct ztest_shell_state){
+		.repeat = 1,
+		.time = true,
+		.sh = sh,
+	};
+
+	ret = ztest_shell_parse_opts(sh, argc, argv, &_state);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = ztest_shell_verify_opts(sh, &_state);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ztest_shell_execute(sh, &_state);
+}
+
+SHELL_CMD_ARG_REGISTER(ztest, NULL, cmd_ztest_help, cmd_ztest, 0, N_ARGS);
+
+bool z_ztest_should_test_run(const char *suite, const char *test)
+{
+	char name[CONFIG_ZTEST_SHELL_NAME_SIZE_MAX];
+	size_t name_size = strlen(suite) + 1 /* '.' */ + strlen(test) + 1 /* '\0' */;
+
+	if (name_size > CONFIG_ZTEST_SHELL_NAME_SIZE_MAX) {
+		printk("Warning: CONFIG_ZTEST_SHELL_NAME_SIZE_MAX is not large enough for\n");
+		printk("Warning: %s.%s\n", suite, test);
+		printk("Warning: Needed: %zu\n", name_size);
+		printk("Warning: Actual: %zu\n", (size_t)CONFIG_ZTEST_SHELL_NAME_SIZE_MAX);
+		return false;
+	}
+
+	snprintf(name, ARRAY_SIZE(name), "%s.%s", suite, test);
+
+	return ztest_shell_filter(name, _state.filter) > 0;
+}
+
+/**
+ * @brief Filter ZTests by name.
+ *
+ * Filter the named @p test (interpreted as `suite.test`) through filters
+ * specified via @p filter. This functionality is inspired by Google Test, where test
+ * executables accept the `--gtest_filter=POSITIVE_PATTERNS[-NEGATIVE_PATTERNS]` CLI option.
+ *
+ * A test should be run (i.e. passes the filter) if it has a name that matches at least one of
+ * `POSITIVE_PATTERNS` but matches precisely zero `NEGATIVE_PATTERNS`. A test should not be run
+ * (i.e. is caught by the filter) if it has a name that matches precisely zero `POSITIVE_PATTERNS`
+ * or at least one `NEGATIVE_PATTERNS`.
+ *
+ * Patterns may include a question mark `?`, which matches a single character, or an asterisk (`*`)
+ * which matches any substring. Multiple patterns may be separated by a colon (`:`).
+ *
+ * @param test The name of the ZTest
+ * @param filter positive and negative patterns
+ *
+ * @retval 0 if the test should not be run
+ * @retval 1 if the test should be run
+ * @retval -EINVAL if an argument is invalid.
+ * @retval -ENOMEM if there is insufficient memory
+ *
+ * @see also tests/unit/ztest/shell/main.c
+ */
+ZTEST_SHELL_STATIC
+int ztest_shell_filter(const char *test, const char *filter)
+{
+	int ret;
+	size_t pmatch_count = 0;
+	/* evaluate matches as positive until a '-' is seen */
+	bool positive_match = true;
+	char pattern[CONFIG_ZTEST_SHELL_PATTERN_MAX];
+
+	if (test == NULL || strlen(test) == 0) {
+		return -EINVAL;
+	}
+
+	if (filter == NULL || strlen(filter) == 0) {
+		return (int)true;
+	}
+
+	/* a sliding window over patterns */
+	bool found_pattern = false;
+
+	pattern[0] = '\0';
+	for (size_t a = 0, b = 0, N = strlen(filter) + 1; a < N && b <= N;) {
```

Comment:
how about having an external variable for `N`? it justs looks odd here

Reply:
That works too.

Reply:
Fixed!